lib/util/cmcd_batching_manager.js

```diff
@@ -0,0 +1,296 @@
+   * @param {!CmcdData} cmcdData The CMCD data to batch
+   * @suppress {checkDebuggerStatement}
+   */
+  addReport(target, cmcdData) {
+    if (this.goneUrls_.has(target.url)) {
+      return;
+    }
+
+    const key = this.generateTargetKey_(target);
+    const serializedCmcd = shaka.util.CmcdUtils.serialize(cmcdData);
+    const urlEncoded = encodeURIComponent(serializedCmcd);
+
+    if (!this.batches_.has(key)) {
+      this.batches_.set(key, {
+        cmcdData: '',
+        target: target,
+      });
+    }
+
+    const batch = this.batches_.get(key);
+    if (batch.cmcdData === '') {
+      batch.cmcdData = urlEncoded;
+    } else {
+      batch.cmcdData += '\n' + urlEncoded;
+    }
+
+    // Set up batch timer if configured and not already set
+    if (target.batchTimer && !this.timers_.has(key)) {
+      const timer = new shaka.util.Timer(() => {
+        this.flushByTargetKey_(key);
+      });
+      timer.tickEvery(target.batchTimer);
+      this.timers_.set(key, timer);
+    }
+
+    // Check if batch size limit is reached
+    if (target.batchSize &&
+        batch.cmcdData.split('\n').length >= target.batchSize) {
+      this.flushByTargetKey_(key);
+    }
+  }
+
+  /**
+   * Flush all batches for a specific URL.
+   *
+   * @param {string} url The URL to flush batches for
+   */
+  flushBatch(url) {
+    for (const key of this.batches_.keys()) {
+      const batch = this.batches_.get(key);
+      if (batch && batch.target && batch.target.url === url) {
+        this.flushByTargetKey_(key);
+      }
+    }
+  }
+
+  /**
+   * Reset the batching manager, clearing all batches and timers.
+   */
+  reset() {
+    for (const timer of this.timers_.values()) {
+      timer.stop();
+    }
+    for (const timer of this.retryTimers_.values()) {
+      timer.stop();
+    }
+    this.batches_.clear();
+    this.timers_.clear();
+    this.retryQueue_ = [];
+    this.retryTimers_.clear();
+    this.goneUrls_.clear();
+  }
+
+  /**
+   * Generate a unique key for a target configuration.
+   *
+   * @param {!shaka.extern.CmcdTarget} target
+   * @return {string}
+   * @private
+   */
+  generateTargetKey_(target) {
+    return JSON.stringify({
+      url: target.url,
+      mode: target.mode,
+      batchSize: target.batchSize,
+      batchTimer: target.batchTimer,
+    });
+  }
+
+  /**
+   * Flush a batch by its target key.
+   *
+   * @param {string} key
+   * @private
+   */
+  flushByTargetKey_(key) {
+    const batch = this.batches_.get(key);
+    if (!batch || batch.cmcdData.length === 0) {
+      return;
+    }
+
+    const {target, cmcdData} = batch;
+    const retryParams = shaka.net.NetworkingEngine.defaultRetryParameters();
+
+    const request = shaka.net.NetworkingEngine.makeRequest(
+        [target.url], retryParams);
+    request.method = 'POST';
+    request.body = shaka.util.StringUtils.toUTF8(cmcdData);
+    request.headers['Content-Type'] = 'application/cmcd+json';
```

Comment:
JSON is no longer supported. Events are now sent as query mode encoded CMCD strings, separated by new lines, with a content type of `text/cmcd`.

See section [3.3](https://docs.google.com/document/d/1isrbeAuauUwjTDUJCxJVltxls_qrFFx7/edit?pli=1#heading=h.2s8eyo1) of the draft spec.